tket/src/WeightSubgrMono/Reducing/DerivedGraphsReducer.cpp

```diff
@@ -0,0 +1,151 @@
+// Copyright 2019-2022 Cambridge Quantum Computing
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "WeightSubgrMono/Reducing/DerivedGraphsReducer.hpp"
+
+#include "Utils/Assert.hpp"
+#include "WeightSubgrMono/Common/SetIntersection.hpp"
+#include "WeightSubgrMono/GraphTheoretic/FilterUtils.hpp"
+#include "WeightSubgrMono/GraphTheoretic/NeighboursData.hpp"
+#include "WeightSubgrMono/Searching/DomainsAccessor.hpp"
+
+namespace tket {
+namespace WeightedSubgraphMonomorphism {
+
+DerivedGraphsReducer::DerivedGraphsReducer(
+    const NeighboursData& pattern_ndata, const NeighboursData& target_ndata)
+    : m_derived_pattern_graphs(
```

Comment:
The SAME objects `m_storage` and `m_counts_storage`, and also `m_calculator`, are passed to both `m_derived_pattern_graphs` and `m_derived_target_graphs` in their constructors.

I.e., they're shared between multiple `DerivedGraphs` objects.

OK, each one could have its own  `m_storage` and `m_counts_storage` object, but what would be the advantage of doing so? (And it would be inefficient in the case of `m_calculator`).

OK, I'll add a comment.

Reply:
Well, OK, I suppose if we just use forward list directly as in the below comment, there's really no need for shared storage; it's identical either way. (But it WOULD make a negligible difference if we started allowing erasures also; but we don't).